torch/csrc/jit/passes/quantization.cpp

```diff
@@ -125,26 +122,56 @@ static void addObserverFor(Value* v, Node* original_observer_node) {
   // Now we can add the inputs.
   observerNode->addInput(v);
   observerNode->addInput(vname);
+  return observerNode;
 }
 
 static bool outputsNeedToBeObserved(Node* n) {
   return n->kind() != prim::Constant;
 }
 
-void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+void InsertObserverNodes(
+    const std::shared_ptr<Graph>& graph,
+    Node* observer_node,
+    size_t num_activation_inputs) {
+  AT_ASSERT(graph != nullptr);
+  // num_activation_inputs is the number of activations or external data
+  // excluding the parameters
+  AT_ASSERT(num_activation_inputs <= graph->inputs().size());
   // For storing all values that need to be instrumented with an observer call.
   std::vector<Value*> values_to_observe;
 
   // For traversing all blocks in the graph including subblocks.
   std::stack<Block*> blocks_to_visit;
 
+  // Mark observer nodes for inputs so we dont add observers
+  // for observers while traversing graph
+  std::unordered_set<Node*> observer_for_input;
+
+  // Add observer for external input nodes excluding parameters
+  // These are treated as activation as they vary across batches
+  // and need to be observed.
+
+  // prim::Param nodes do not belong to the graph. Hence the Insert
+  // point is the beginning of graph node. This also safe guards against
+  // observing a potentially mutated value due to some in-place operation
+  Node* insert_node = *graph->nodes().begin();
+  for (size_t idx = 0; idx < num_activation_inputs; ++idx) {
+    auto& v = graph->inputs()[idx];
+    if (v->type()->isSubtypeOf(TensorType::get())) {
+      Node* new_observer_node = addObserverFor(v, observer_node, insert_node);
+      new_observer_node->insertBefore(insert_node);
+      observer_for_input.emplace(new_observer_node);
+    }
+  }
+
```

Comment:
Since this function now has two logical parts, please add a comment for the second part too.